Tools/clinic/clinic.py

```diff
@@ -1915,8 +1916,11 @@ def dump(self):
 # maps strings to Language objects.
 # "languages" maps the name of the language ("C", "Python").
 # "extensions" maps the file extension ("c", "py").
+L = TypeVar('L', bound=Language)
+LangDict = dict[str, L]
```

Comment:
I'm slightly surprised I cannot just do this:

```python
LangDict = dict[str, Language]
```

Reply:
What's the error you get if you do that?

Reply:
```
Tools/clinic/clinic.py:1922: error: Value expression in dictionary comprehension has incompatible type "Type[CLanguage]"; expected type "Language"  [misc]
Tools/clinic/clinic.py:1923: error: Incompatible types in assignment (expression has type "Type[PythonLanguage]", target has type "Language")  [assignment]
Found 2 errors in 1 file (checked 1 source file)
```

Reply:
Right, so it should be LangDict = dict[str, type[Language]]` rather than `dict[str, Language]` or `dict[str, L]`

Reply:
Aaaaaah #til

Reply:
Nah, I get this now:

```
Tools/clinic/clinic.py:1931: error: Only concrete class can be given where "Type[Language]" is expected  [type-abstract]
```

Reply:
Hmm, you could try `dict[str, Callable[[str], Language]]` instead (all class objects can also be thought of as callables that return instances of that class).

Or tbh you could just run mypy with `--disable-error-code="type-abstract"` -- this is probably the least useful lint mypy provides, and is quite controversial.

Reply:
For some heated debate on whether this error code is useful at all, you can see https://github.com/python/mypy/issues/4717

Reply:
Seems like this is a bug in mypy (Language is a class with an abc metaclass)

Reply:
I'll just revert to what I had earlier

Reply:
Hmm, what you had earlier is incorrect :/

Mypy is working as intended here -- it's a question of the design choice mypy has made here rather than a _bug_, per se. Mypy believes that if you have a `type[Foo]` annotation, you'll probably want to instantiate `Foo` at some point, so it's therefore unsound to have a `type[Foo]` annotation if `Foo` is an abstract class, since abstract classes can't be instantiated. In this case `Language` is an abstract class (it has an abc metaclass and abstract methods, so it can't be instantiated). So mypy is working as designed.

However, arguably this lint is based on some flawed assumptions and makes life unduly complicated in lots of situations.

You can work around the issue by using the `Callable[[str], Language]` annotation I suggested above, or disable the error code. But using a `TypeVar` is incorrect here, since there's no type-dependent relationship going on.

Reply:
Got it; I'll investigate tomorrow. (FYI, your comments did not show up until now.)